### PR TITLE
make ConfigMap collector namespace-optional

### LINKED ISSUE #907: ConfigMap collector should be namespace optional
**Describe the rationale for the suggested feature.**

The ConfigMap collector spec requires namespace to be provided.  As a vendor, we try not to dictate our installation namespace, so we don't know that ahead of time. If we hardcode a namespace into the support-bundle.yaml, we will not collect that CM if the namespaces don't match.

Some other collectors are already namespace-optional, for example `logs`, which makes sense:

    [namespace (Optional)](https://troubleshoot.sh/docs/collect/logs/#namespace-optional)
    The namespace to search for the pod selector in. If this is not provided, it will default to the current namespace of the context.

**Describe the feature**

The ConfigMap collector should parallel the logs one here.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by adamancini**
## Description, Motivation and Context

When no namespace is given in the ConfigMaps collector spec, use the namespace from the current kubectl context.

Fixes: #907 

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [x] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by adamancini** [comment]
Using the test spec from my PR:

```
tail support-bundle-2023-06-06T10_09_37/configmaps/* support-bundle-2023-06-06T10_09_37/configmaps-errors/* 
==> support-bundle-2023-06-06T10_09_37/configmaps/cluster-info.json <==
{
  "namespace": "",
  "name": "cluster-info",
  "key": "",
  "configMapExists": false,
  "keyExists": false,
  "data": null
}
==> support-bundle-2023-06-06T10_09_37/configmaps/does-not-exist.json <==
{
  "namespace": "",
  "name": "does-not-exist",
  "key": "",
  "configMapExists": false,
  "keyExists": false,
  "data": null
}
==> support-bundle-2023-06-06T10_09_37/configmaps/kube-proxy.json <==
{
  "namespace": "",
  "name": "kube-proxy",
  "key": "",
  "configMapExists": true,
  "keyExists": false,
  "data": null
}
==> support-bundle-2023-06-06T10_09_37/configmaps/kube-root-ca.crt.json <==
{
  "namespace": "",
  "name": "kube-root-ca.crt",
  "key": "",
  "configMapExists": true,
  "keyExists": false,
  "data": null
}
==> support-bundle-2023-06-06T10_09_37/configmaps/kube-system <==

==> support-bundle-2023-06-06T10_09_37/configmaps-errors/cluster-info.json <==
[
  "configmaps \"cluster-info\" not found"
]
==> support-bundle-2023-06-06T10_09_37/configmaps-errors/does-not-exist.json <==
[
  "configmaps \"does-not-exist\" not found"
]%                                                                                                   
```

File: pkg/collect/configmap.go [hunk 1]
```diff
@@ -49,8 +51,20 @@ func (c *CollectConfigMap) Collect(progressChan chan<- interface{}) (CollectorRe
...
 
 	configMaps := []corev1.ConfigMap{}
+	namespace := c.Collector.Namespace
+
 	if c.Collector.Name != "" {
-		configMap, err := c.Client.CoreV1().ConfigMaps(c.Collector.Namespace).Get(c.Context, c.Collector.Name, metav1.GetOptions{})
+		if namespace == "" {
+			kubeconfig := k8sutil.GetKubeconfig()
+			ns, _, err := kubeconfig.Namespace()
+			klog.V(2).Infof("no namespace was set for configmap '%s': using namespace '%s' from current kubectl context", c.Collector.Name, ns)
```
Comment (diamonwiggins) [inline pkg/collect/configmap.go]:
```suggestion
			klog.V(2).Infof("no namespace was set for configmap '%s': using namespace '%s' from current kubeconfig context", c.Collector.Name, ns)
```

**Review by diamonwiggins — APPROVED** [review]
Approved with a suggestion, and a caveat of there has been some discussion recently if we should ever use the `kubeconfig` context's namespace at all. Not sure what the argument is there, but i don't object to this change.

**Review by diamonwiggins — DISMISSED** [review]
Approved with a suggestion, and a caveat of there has been some discussion recently if we should ever use the `kubeconfig` context's namespace at all. Not sure what the argument is there, but i don't object to this change.

**Review by diamonwiggins — APPROVED** [review]

File: pkg/collect/configmap.go [hunk 2]
```diff
@@ -49,8 +51,20 @@ func (c *CollectConfigMap) Collect(progressChan chan<- interface{}) (CollectorRe
...
+		if namespace == "" {
+			kubeconfig := k8sutil.GetKubeconfig()
+			ns, _, err := kubeconfig.Namespace()
+			klog.V(2).Infof("no namespace was set for configmap '%s': using namespace '%s' from current kubeconfig context", c.Collector.Name, ns)
+			if err != nil {
+				return nil, errors.Wrapf(err, "a namespace was not specified for configmap '%s' and could not be discovered from kubeconfig", c.Collector.Name)
+			}
+			namespace = ns
+		}
+		klog.V(1).Infof("looking for configmap '%s' in namespace '%s'", c.Collector.Name, namespace)
```
Comment (banjoh) [inline pkg/collect/configmap.go:66]: Within the collectors, this needs to be `klog.V(2)` as per [this "very easy to find" documentation](https://github.com/replicatedhq/troubleshoot/blob/a5628874b08c76a8544dfba6da62c7605118d91f/pkg/logger/logger.go#L4)

Note to self: I really need to extract this into our [dev guide documentation](https://github.com/replicatedhq/troubleshoot/issues/1031)

**Review by banjoh — CHANGES_REQUESTED** [review]

**Review by banjoh — APPROVED** [review]

**Pull request merged**